### PR TITLE
Bind, Unbind and validations

### PR DESCRIPTION
**Pull request by MarinhoFeliphe**
# Bind object definitions:

Take a look at the example below, each circle is an object definition and each arrow is a relationship, red object definitions are inside a hierarchical structure where the root is the object definition A.

image

the MVC Resource Command below is responsible to bind object definitions based on the edges provided (Path),

**p_p_resource_id** - `/object_definitions/bind_object_definitions`

## Parameter:

- `objectRelationshipIds`: here you need a list of edges (object relationship ids) where the last item is the root, the penult is the edge depth 1, antepenult is the edge depth 2 and so on.

## Examples:

On frontend side we can't create the Tree above executing one bind request, in this case will be need at leas 3 requests, the request below will create this structure:

**Request Parameters:** `objectRelationshipIds=[edge3Id, edge1Id]`

image

to add the object definition AAB and AB you need to perform the following requests:

**Bind AAB Request Parameters:** `objectRelationshipIds=[edge4Id, edge1Id]`

**Bind AB Request Parameters:** `objectRelationshipIds=[edge2Id]`

It's like a path of the position where that node should be bind.

# Unbind object definitions:

the MVC Resource Command below is responsible to unbind an object definition and its descendants,

**p_p_resource_id** - `/object_definitions/unbind_object_definition`

## Parameters:

- `objectDefinitionId`: pass the object definition id you wanna unbind

### Unbind a leaf:

Unbind AB request Parameter: `objectDefinitionId=ABId`

**Result:**

image

### Unbind an internal node:

Unbind AA request Parameter: `objectDefinitionId=ABId`

**Result:**

image

### CODE REVIEW

File: modules/apps/object/object-test/src/testIntegration/java/com/liferay/object/web/internal/portlet/action/test/UnbindObjectDefinitionMVCResourceCommandTest.java [hunk 1]
```diff
@@ -0,0 +1,313 @@
...
+					_objectDefinitionLocalService.fetchObjectDefinition(
+						node.getObjectDefinitionId());
+
+				sb.append(objectDefinition.getShortName());
+			});
+
+		return sb.toString();
+	}
+
+	private void _unbindObjectDefinition(long objectDefinitionId)
```
Comment (gabrielwas) [inline modules/apps/object/object-test/src/testIntegration/java/com/liferay/object/web/internal/portlet/action/test/UnbindObjectDefinitionMVCResourceCommandTest.java:258]: @MarinhoFeliphe rename to `_unbind`

**Review by gabrielwas — CHANGES_REQUESTED** [review]
@MarinhoFeliphe Check the comments, please.

I didn't fully review the following files, since we had several changes in the original PR:
`UnbindObjectDefinitionMVCResourceCommandTest`
`BindObjectDefinitionsMVCResourceCommandTest`

File: modules/apps/object/object-test/src/testIntegration/java/com/liferay/object/web/internal/portlet/action/test/UnbindObjectDefinitionMVCResourceCommandTest.java (hunk 1 above)
Comment (MarinhoFeliphe) [reply modules/apps/object/object-test/src/testIntegration/java/com/liferay/object/web/internal/portlet/action/test/UnbindObjectDefinitionMVCResourceCommandTest.java]: done [here](https://github.com/liferay-objects/liferay-portal/pull/2458/commits/0daa57981a1e5533cbc4d4cbe74be7ee030cd1f9)

**Comment by MarinhoFeliphe** [comment]
ci:test:relevant

**Comment by MarinhoFeliphe** [comment]
ci:test:sf

**Comment by MarinhoFeliphe** [comment]
Hi @carolmariaabb,

just to make sure that this went unnoticed, as Gabriel said [here](https://github.com/liferay-objects/liferay-portal/pull/2458#pullrequestreview-1593822228) he reviewed all other changes but these one below:

- UnbindObjectDefinitionMVCResourceCommandTest
- BindObjectDefinitionsMVCResourceCommandTest

so feel free to decide review only these two tests or the whole pull request,

Thanks!

File: modules/apps/object/object-service/src/main/java/com/liferay/object/internal/definition/tree/TreeFactoryImpl.java [hunk 2]
```diff
@@ -27,7 +30,15 @@
...
 
 	@Override
-	public Tree create(long objectDefinitionId) {
+	public Tree create(long objectDefinitionId) throws PortalException {
+		ObjectDefinition objectDefinition =
+			_objectDefinitionLocalService.getObjectDefinition(
+				objectDefinitionId);
+
+		if (objectDefinition.getRootObjectDefinitionId() == 0) {
+			return null;
```
Comment (carolmariaabb) [inline modules/apps/object/object-service/src/main/java/com/liferay/object/internal/definition/tree/TreeFactoryImpl.java:39]: @MarinhoFeliphe what do you think about throwing an exception here? 

So when we use `TreeFactory.create` we wouldn't need to handle the exception AND check if `Tree` is `null`.

File: modules/apps/object/object-web/src/main/java/com/liferay/object/web/internal/object/definitions/portlet/action/UnbindObjectDefinitionMVCResourceCommand.java [hunk 3]
```diff
@@ -0,0 +1,92 @@
...
+
+		Iterator<Node> iterator = tree.iterator(
+			objectDefinition.getObjectDefinitionId());
+
+		while (iterator.hasNext()) {
+			_unbindNode(iterator.next());
+		}
+	}
+
+	private void _unbindNode(Node node) throws Exception {
```
Comment (carolmariaabb) [inline modules/apps/object/object-web/src/main/java/com/liferay/object/web/internal/object/definitions/portlet/action/UnbindObjectDefinitionMVCResourceCommand.java]: @MarinhoFeliphe do we really need this private method? It's only used once.

**Review by carolmariaabb — COMMENTED** [review]
@MarinhoFeliphe, great job 🎉 

I made a few suggestions, s, but I don't see a problem with forwarding the PR without them.

File: modules/apps/object/object-test/src/testIntegration/java/com/liferay/object/web/internal/portlet/action/test/UnbindObjectDefinitionMVCResourceCommandTest.java [hunk 4]
```diff
@@ -0,0 +1,158 @@
...
+
+		themeDisplay.setLocale(LocaleUtil.getSiteDefault());
+		themeDisplay.setTimeZone(TimeZoneUtil.getDefault());
+
+		mockLiferayResourceRequest.setAttribute(
+			WebKeys.THEME_DISPLAY, themeDisplay);
+
+		mockLiferayResourceRequest.addParameter(
+			"objectDefinitionId",
+			String.valueOf(objectDefinition.getObjectDefinitionId()));
```
[the remark this answers is not in the document]
Comment (MarinhoFeliphe) [reply modules/apps/object/object-test/src/testIntegration/java/com/liferay/object/web/internal/portlet/action/test/UnbindObjectDefinitionMVCResourceCommandTest.java]: Done [here](https://github.com/liferay-objects/liferay-portal/pull/2458/commits/187546ceece090366a010c7ad5e3883a8a163b05)

File: modules/apps/object/object-test/src/testIntegration/java/com/liferay/object/service/test/util/TreeTestUtil.java [hunk 5]
```diff
@@ -14,55 +14,125 @@
...
+
+		ThemeDisplay themeDisplay = new ThemeDisplay();
+
+		themeDisplay.setLocale(LocaleUtil.getSiteDefault());
+		themeDisplay.setTimeZone(TimeZoneUtil.getDefault());
+
+		mockLiferayResourceRequest.setAttribute(
+			WebKeys.THEME_DISPLAY, themeDisplay);
+
+		mockLiferayResourceRequest.addParameter(
```
[the remark this answers is not in the document]
Comment (MarinhoFeliphe) [reply modules/apps/object/object-test/src/testIntegration/java/com/liferay/object/service/test/util/TreeTestUtil.java]: Done [here](https://github.com/liferay-objects/liferay-portal/pull/2458/commits/239763350fc871b331ee5ef9cd3fc091b0935be3)

File: modules/apps/object/object-service/src/main/java/com/liferay/object/internal/definition/tree/TreeFactoryImpl.java (hunk 2 above)
Comment (MarinhoFeliphe) [reply modules/apps/object/object-service/src/main/java/com/liferay/object/internal/definition/tree/TreeFactoryImpl.java]: Agreed, good catch! done [here](https://github.com/liferay-objects/liferay-portal/pull/2458/commits/a5baa33e3c3adfdba69be2dc8391917529668aec)

File: modules/apps/object/object-web/src/main/java/com/liferay/object/web/internal/object/definitions/portlet/action/UnbindObjectDefinitionMVCResourceCommand.java (hunk 3 above)
Comment (MarinhoFeliphe) [reply modules/apps/object/object-web/src/main/java/com/liferay/object/web/internal/object/definitions/portlet/action/UnbindObjectDefinitionMVCResourceCommand.java]: Done [here](https://github.com/liferay-objects/liferay-portal/pull/2458/commits/71af23a1e052720208238914c7574f2a99517a87)

File: modules/apps/object/object-test/src/testIntegration/java/com/liferay/object/service/test/util/TreeTestUtil.java [hunk 6]
```diff
@@ -14,55 +14,125 @@
...
+import javax.portlet.PortletException;
 
 /**
  * @author Feliphe Marinho
  */
 public class TreeTestUtil {
 
+	public static void assertTree(
+			Map<String, String[]> expectedMap, Tree actualTree,
+			ObjectDefinitionLocalService objectDefinitionLocalService)
```
[the remark this answers is not in the document]
Comment (MarinhoFeliphe) [reply modules/apps/object/object-test/src/testIntegration/java/com/liferay/object/service/test/util/TreeTestUtil.java:41]: Hi @carolmariaabb I've followed the same pattern used in Assert's method, the expected parameter comes before the actual parameter, [for example](https://github.com/liferay/liferay-portal/blob/6ce0b9eb049e99a50285a324a3bd4afd76d340ab/portal-test/src/com/liferay/portal/kernel/test/AssertUtils.java#L67),

I'm not sure about the place of objectDefinitionLocalService, let me know if makes sense to you!

maybe I should put this parameter before the expectedMap

**Comment by MarinhoFeliphe** [comment]
hey @carolmariaabb thanks for your accurate review,

please see my comments,

Thank you!

**Review by carolmariaabb — APPROVED** [review]
LGTM 🎉

**Comment by MarinhoFeliphe** [comment]
ci:forward

**Comment by Luizcarlosqueiroz** [comment]
ci:test:objects

**Comment by Luizcarlosqueiroz** [comment]
ci:test:object

**Comment by MarinhoFeliphe** [comment]
Resent here: https://github.com/brianchandotcom/liferay-portal/pull/139919

**Pull request closed without merging**